File: 2021/kotlin/src/main/kotlin/nl/sanderp/aoc/aoc2021/day04/Day04.kt
package nl.sanderp.aoc.aoc2021.day04

import nl.sanderp.aoc.common.measureDuration
import nl.sanderp.aoc.common.prettyPrint
import nl.sanderp.aoc.common.readResource

fun main() {
    val (numbers, boards) = parse(readResource("Day04.txt").lines())

    val (answer1, duration1) = measureDuration<Int> { partOne(numbers, boards) }
    println("Part one: $answer1 (took ${duration1.prettyPrint()})")

    val (answer2, duration2) = measureDuration<Int> { partTwo(numbers, boards) }
    println("Part two: $answer2 (took ${duration2.prettyPrint()})")
}

private fun parse(input: List<String>): Pair<List<Int>, List<Board>> {
    val numbers = input.first().split(",").map { it.toInt() }
    val boards = input.drop(1).chunked(6).map { lines ->
        Board(lines.drop(1).map { l -> l.split(" ").filterNot { it == "" }.map { it.toInt() } })
    }

    return numbers to boards
}

class Board(numbers: List<List<Int>>) {
    private val rows = numbers.map { row -> row.map { it to false }.toTypedArray() }.toTypedArray()
    private val cols
        get() = rows.indices.map { i -> rows.indices.map { j -> rows[j][i] } }

    val unmarked
        get() = rows.flatMap { row -> row.filterNot { it.second } }.map { it.first }

    fun mark(number: Int): Boolean {
        for (i in rows.indices) {
            for (j in rows[i].indices) {
                if (rows[i][j].first == number) {
                    rows[i][j] = number to true
                }
            }
        }

        return rows.any { row -> row.all { it.second } } || cols.any { col -> col.all { it.second }}
    }
}

private fun partOne(numbers: List<Int>, boards: List<Board>): Int {
    for (number in numbers) {
        for (board in boards) {
            if (board.mark(number)) {
                return board.unmarked.sum() * number
            }
        }
    }

    return -1
}

private fun partTwo(numbers: List<Int>, boards: List<Board>): Int {
    val playing = boards.toMutableList()

    for (number in numbers) {
        val wonRound = mutableListOf<Board>()
        for (board in playing) {
            if (board.mark(number)) {
                if (playing.size - wonRound.size == 1) {
                    return board.unmarked.sum() * number
                }

                wonRound.add(board)
            }
        }
        playing.removeAll(wonRound)
    }

    return -1
}

File: 2022/scala/src/main/scala/AdventOfCode2022/Day04.scala
package AdventOfCode2022

object Day04 {
  def parse(input: String): (Seq[Int], Seq[Int]) = {
    val Array(first, second) = input.split(',').map { case s"$lo-$hi" => lo.toInt to hi.toInt }
    (first, second)
  }

  def partOne(input: Seq[String]): Int = input
    .map(parse)
    .count(x => x._1.containsSlice(x._2) || x._2.containsSlice(x._1))

  def partTwo(input: Seq[String]): Int = input
    .map(parse)
    .count(x => x._1.intersect(x._2).nonEmpty)

  def main(args: Array[String]): Unit = {
    val input = io.Source.fromResource("Day04.txt").getLines().toSeq
    println(s"Part one: ${partOne(input)}")
    println(s"Part two: ${partTwo(input)}")
  }
}
